### PR TITLE
targetLinkLibrariesWithDynamicLookup: Fix support for cross-compilation exception

### PR DESCRIPTION
**Pull request by erykoff**
This pull-request fixes dynamic linking when cross-compiling on macOS.

It revisits the approach used to support cross-compilation when no emulator is available.

This is done by further improving on fixes integrated through https://github.com/scikit-build/scikit-build/commit/c0ec848644a22fb165018be7882b24099dcec2ae (`Fix targetLinkLibrariesWithDynamicLookup`) and https://github.com/scikit-build/scikit-build/commit/07b2c4ff0b03f830456eb1c17bae87437d5bf5c5 (`fix: assume success for osx_dynamic_lookup when cross compiling`)

Messages reported during configuration are also improved, when the compile tests are assumed to pass when cross-compiling without emulator, a clear message is reported.

For example:

```
Performing Test Weak Link MODULE -> SHARED (osx_dynamic_lookup) - Assuming Success (Cross compiling without emulator on macOS)
```

Fixes:
* #900
* https://github.com/conda-forge/h3-py-feedstock/issues/38

<details><summary>Previous description</summary>
This PR is a possible solution to #900, building on what @isuruf had done previously.  I don't think this would have any bad side effects, but I'm not very familiar with cmake so I could be wrong.

I did confirm that with this patch then `h3-py` cross-compiles correctly, so this would fix https://github.com/conda-forge/h3-py-feedstock/issues/38
</details>

### CODE REVIEW

**Comment by henryiii** [comment]
I think this is reasonable. @jcfr or @aaron-bray, let me know if you see any problems, otherwise I think it's safe to merge.

**Comment by erykoff** [comment]
Thanks!  This does look cleaner than my hacked together patch.  I will test in my use case and report back.

File: skbuild/resources/cmake/targetLinkLibrariesWithDynamicLookup.cmake [hunk 1]
```diff
@@ -198,6 +198,28 @@ function(_test_weak_link_project
...
 
+  if(CMAKE_CROSSCOMPILING)
+    set(link_flag_spec "no_flag")
+    set(link_flag "${link_flag_spec}")
+    set(test_skipping_reason "")
+    set(test_pass FALSE)
+
+    if(APPLE AND NOT CMAKE_CROSSCOMPILING_EMULATOR)
+      set(link_flag_spec "osx_dynamic_lookup")
+      set(link_flag "${link_flag_spec}")
```
Comment (erykoff) [inline skbuild/resources/cmake/targetLinkLibrariesWithDynamicLookup.cmake:209]: So this doesn't work, the cross-build errors with `error: no such file or directory: 'osx_dynamic_lookup'`.  I think that there's something here with a string replacement that isn't happening (it should be substituting `-undefined dynamic_lookup`) but something with quotes/brackets/?? is going wrong.

Comment (jcfr) [reply skbuild/resources/cmake/targetLinkLibrariesWithDynamicLookup.cmake]: Thanks for testing, the next commit should address this.

_Note that once we have confirmed it works, commits will be squashed_

**Comment by erykoff** [comment]
With the latest fix, it works for my test cross compiling from osx-64 to osx-arm64.  Thanks!

**Review by henryiii — APPROVED** [review]

**Comment by jcfr** [comment]
> With the latest fix, it works for my test cross compiling from osx-64 to osx-arm64. Thanks!

Great. I will update the commit message and this should be good for integration.

**Comment by henryiii** [comment]
When you squash and merge, you can set the commit message there. ;)

**Pull request merged**